How many late days are you using for this assignment? 

### 1 (Submitted 3/8/2025)

In [156]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [157]:
# Load the given datasets
catdata = pd.read_csv(r"C:\Users\sagea\OneDrive\CS 383\chronic_kidney_disease_categorical.csv")
numdata = pd.read_csv(r"C:\Users\sagea\OneDrive\CS 383\chronic_kidney_disease_numerical.csv")

###Comment: I am using Jupyter notebook navigator on the web, so I had to look up how to access my CSV. The instructions were unclear. 
# Print the data

print("Below is a snippet of the categorical data:\n" ,catdata)
print("\n")
print("Below is a snippet of the numerical data:\n", numdata)

Below is a snippet of the categorical data:
      unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane 

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Note: This is a very oddly worded question... ("what THE each data IS"...?, what is "each data", like every column?... I am just going to describe THE data, forgive me if that is not what was intended.)

Answer: The data seems to be representing health data of a large set of individuals some with and some without Chronic Kidney disease. The features are things such as age that give information about each patient that allows us to draw insights about different aspects of their health that may or may not be affected or have an effect on their Chronic Kidney disease, and their labels are the numerical or categorial representation the amount of type of feature the certain patient has. Since we are looking at patients with both CKD and notCKD it can allow us to find potential correlations between a patient's features and the presence of CKD. The features in the data set are both numerical and categorial.  3 examples of numerical columns are age, blood pressure, and white blood cell count. The categorials are: specific gravity, 	albumi, and -	sugar

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [158]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print("Numerical Data:\n)")
numduplicates = numdata.duplicated(keep=False)
print(numduplicates)
# Drop these duplicate rows
numdata_nodups= numdata.drop_duplicates()
print(numdata_nodups)
# Number of duplicate rows
print("30")

### Was there a better way to do this besides looking at the amount of rows in dropped duplicates, then subtracting that from the amount of rows in the original?
### Like is there a function that just prints the number?
print("The original numerical data set had 420, and the numerical data with the dropped duplicates have 390, so there were 30 sets of duplicated data")
print("\n\n\n\n\n\n\n\n")

print("Categorial Data:\n)")
# Repeat the same for categorical dataset. Print the duplicate rows and drop them
catduplicates = catdata.duplicated(keep=False)
print(catduplicates)
# Drop these duplicate rows
catdata_nodups = catdata.drop_duplicates()
print(catdata_nodups)
# Number of duplicate rows
print("30")

print("The original numerical data set had 430, and the categorial data with the dropped duplicates have 400, so there were 30 sets of duplicated data")



Numerical Data:
)
0      False
1       True
2      False
3      False
4       True
       ...  
415    False
416    False
417    False
418    False
419    False
Length: 420, dtype: bool
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: I do not think that both of the rows in the data set are present in both sets, mainly because the categorial data has more rows than the numerical data. I believe that the rows in the combined table will have empty spaces in it depending on the from which set they originate from.

In [159]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combineddata = pd.merge(numdata,catdata, on = 'unique_id')
combineddata_nodups = pd.merge(numdata_nodups, catdata_nodups, on = 'unique_id')
#Print the combined dataset
print(combineddata)

print('\n\n\n\n\n\n\n\nAttached is the combined data set:\n', combineddata_nodups)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
3       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
4       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
477     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
478     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
479     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
480     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
481     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc          ba  htn   dm cad

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [160]:
# Calculate the percentage of rows that contain atleast one missing value
amount_of_rows_with_missing = combineddata[combineddata.isnull().any(axis=1)]
print(amount_of_rows_with_missing)
# Print %
print((284/482)*100, '% of the rows from the combined data set are missing a point of data')
# Drop these rows from the dataset
combineddata_nomissing = combineddata.dropna()
# Print the Dataset

print('\n\n\n\n', combineddata_nomissing)

     unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
4       764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
5       240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
6       240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
7       992643  35.0  70.0   82.0   36.0  1.1  150.0  3.5  14.5  52.0  ...   
8       573400  60.0  80.0    NaN    NaN  NaN    NaN  NaN   NaN   NaN  ...   
..         ...   ...   ...    ...    ...  ...    ...  ...   ...   ...  ...   
474     158929  73.0   NaN   95.0   51.0  1.6  142.0  3.5   NaN   NaN  ...   
475     158929  73.0   NaN   95.0   51.0  1.6  142.0  3.5   NaN   NaN  ...   
476     449458  54.0  80.0  207.0   77.0  6.3  134.0  4.8   9.7  28.0  ...   
477     522912  56.0  80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
481     926033  50.0  70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc          ba  htn   dm cad appet   pe 

## Part 3.4: Sort the dataset according to the Labels

In [161]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

# Print the dataset
ncombdata = combineddata_nomissing
sortedncombdata = ncombdata.sort_values(by='Target').reset_index()
print('Combined Data Set Sorted by Target:\n', sortedncombdata)

Combined Data Set Sorted by Target:
      index  unique_id   age     bp    bgr     bu   sc    sod  pot  hemo  ...  \
0      480     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  ...   
1      105     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  ...   
2      118     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  ...   
3      397     337834  48.0   70.0  117.0   56.0  3.8  111.0  2.5  11.2  ...   
4      396     337834  48.0   70.0  117.0   56.0  3.8  111.0  2.5  11.2  ...   
..     ...        ...   ...    ...    ...    ...  ...    ...  ...   ...  ...   
193    204     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  ...   
194    205     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  ...   
195    208     261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  ...   
196    273     489160  28.0   60.0   79.0   50.0  0.5  145.0  5.0  17.6  ...   
197    247     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  ...   

  

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [162]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
ncombenc = pd.get_dummies(sortedncombdata, columns=categorical_columns, drop_first=False, dtype= int)

print(ncombenc)
# Print the dataset

     index  unique_id   age     bp    bgr     bu   sc    sod  pot  hemo  ...  \
0      480     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  ...   
1      105     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  ...   
2      118     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  ...   
3      397     337834  48.0   70.0  117.0   56.0  3.8  111.0  2.5  11.2  ...   
4      396     337834  48.0   70.0  117.0   56.0  3.8  111.0  2.5  11.2  ...   
..     ...        ...   ...    ...    ...    ...  ...    ...  ...   ...  ...   
193    204     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  ...   
194    205     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  ...   
195    208     261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  ...   
196    273     489160  28.0   60.0   79.0   50.0  0.5  145.0  5.0  17.6  ...   
197    247     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  ...   

     cad_no  cad_yes  appet_good  appet

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: In the downstream modeling this could be equivalent, because depending on the data category you would like to sort, you could do a similar thing than what we did in Part 3.4. If the categorial variable took more than 2 values this wouldn't work in binary encoding, however it could possibly work using 5 numerical values and sorting by that instead however it could me more inefficent because you'd need to individually sort by each "condiment" (value 1-5) instead of just sorting observing which sandwiches are in column condiment_ketchup... for example. 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [163]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
### I used an online source to help me figure out how to do this...
for col in numerical_columns:
    mean = ncombenc[col].mean()
    std_dev = ncombenc[col].std()
    outliers_condition = (ncombenc[col] < mean - 3 * std_dev) | (ncombenc[col] > mean + 3 * std_dev)
    ncombenc_nooutliers = ncombenc[~outliers_condition]
print(ncombenc_nooutliers)

     index  unique_id   age     bp    bgr     bu   sc    sod  pot  hemo  ...  \
0      480     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  ...   
1      105     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  ...   
2      118     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  ...   
3      397     337834  48.0   70.0  117.0   56.0  3.8  111.0  2.5  11.2  ...   
4      396     337834  48.0   70.0  117.0   56.0  3.8  111.0  2.5  11.2  ...   
..     ...        ...   ...    ...    ...    ...  ...    ...  ...   ...  ...   
193    204     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  ...   
194    205     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  ...   
195    208     261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  ...   
196    273     489160  28.0   60.0   79.0   50.0  0.5  145.0  5.0  17.6  ...   
197    247     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  ...   

     cad_no  cad_yes  appet_good  appet

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [164]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
ncomb_norm = ncombenc_nooutliers
for col in numerical_columns:
    min = ncomb_norm [col].min()
    max = ncomb_norm [col].max()
    ncomb_norm[col] = (ncomb_norm[col] - min) / (max-min)
    ### I had to search up what "normalized data" was, and found this metric. 
print(ncomb_norm)
# Print the dataset

     index  unique_id       age        bp       bgr        bu        sc  \
0      480     474407  0.194805  0.666667  0.088095  0.100334  0.087838   
1      105     137148  0.870130  0.833333  0.535714  0.267559  0.351351   
2      118     484175  0.610390  0.666667  0.000000  0.324415  0.459459   
3      397     337834  0.545455  0.333333  0.111905  0.153846  0.229730   
4      396     337834  0.545455  0.333333  0.111905  0.153846  0.229730   
..     ...        ...       ...       ...       ...       ...       ...   
193    204     769512  0.415584  0.166667  0.050000  0.086957  0.020270   
194    205     769512  0.415584  0.166667  0.050000  0.086957  0.020270   
195    208     261137  0.376623  0.166667  0.083333  0.096990  0.006757   
196    273     489160  0.285714  0.166667  0.021429  0.133779  0.006757   
197    247     572140  0.636364  0.500000  0.080952  0.060201  0.033784   

          sod       pot      hemo  ...  cad_no  cad_yes  appet_good  \
0    0.358974  0.022472  0.3

C:\Users\sagea\AppData\Local\Temp\ipykernel_27284\2221924952.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ncomb_norm[col] = (ncomb_norm[col] - min) / (max-min)


#### MY CODE IS RUNNING FINE... THIS ERROR POPPED UP AFTER I COMPLETED THE ASSIGNMENT. 

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Below I checked to see if the ane and the pe columns are necessary because from the snippets that I see above, they look all uniform. Upon closer inspection, by dropping duplicates in both columns, I can see that there are rows in both columns that contain different values so they are appropriate for modeling because it is a metric that can model can use to filter. The only other column that I don't think would be necessary is the original index column "Index" because now we have a NEW indices sorted by the "Target" value. 

In [165]:
newN = ncomb_norm 
newNay = newN['ane_yes'].drop_duplicates()
print(newNay)
newNpn = newN['pe_no'].drop_duplicates()
print(newNpn)

### The lines above are checking to see if two of the categorial columns are necessary, since they contain different values then they must be. 
print('\n\n') 
FinalData = newN.drop('index', axis=1)
print(FinalData)

0    1
1    0
Name: ane_yes, dtype: int32
0    1
3    0
Name: pe_no, dtype: int32



     unique_id       age        bp       bgr        bu        sc       sod  \
0       474407  0.194805  0.666667  0.088095  0.100334  0.087838  0.358974   
1       137148  0.870130  0.833333  0.535714  0.267559  0.351351  0.743590   
2       484175  0.610390  0.666667  0.000000  0.324415  0.459459  0.076923   
3       337834  0.545455  0.333333  0.111905  0.153846  0.229730  0.000000   
4       337834  0.545455  0.333333  0.111905  0.153846  0.229730  0.000000   
..         ...       ...       ...       ...       ...       ...       ...   
193     769512  0.415584  0.166667  0.050000  0.086957  0.020270  0.615385   
194     769512  0.415584  0.166667  0.050000  0.086957  0.020270  0.615385   
195     261137  0.376623  0.166667  0.083333  0.096990  0.006757  0.615385   
196     489160  0.285714  0.166667  0.021429  0.133779  0.006757  0.871795   
197     572140  0.636364  0.500000  0.080952  0.060201  0

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [166]:
# Export the dataframe to a new csv file
FinalData.to_csv("cleanedData.csv")
# Export the dataframe to a new json file
FinalData.to_json("cleanedData.json")

### is this line of code meant to do anything?, cause when I run it is just sets. 

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [167]:
print(combineddata_nomissing.head(15))

    unique_id   age    bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0      203694  38.0  80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1      938027  43.0  60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2      938027  43.0  60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
3      421471  37.0  60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
9      109053  64.0  60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
15     349892  59.0  70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
20     630177  58.0  80.0  131.0  18.0  1.1  141.0  3.5  15.8  53.0  ...   
21     741276  23.0  60.0   95.0  24.0  0.8  145.0  5.0  15.0  52.0  ...   
22     546010  49.0  80.0  122.0  32.0  1.2  139.0  3.9  17.0  41.0  ...   
23     460465  47.0  60.0  109.0  25.0  1.1  141.0  4.7  15.8  41.0  ...   
25     965166  48.0  80.0   75.0  22.0  0.8  137.0  5.0  16.8  51.0  ...   
26     201677  33.0  80.0   89.0  19.0  1.1  144.0  5.0  15.0  40.0  ...   
27     97687

Paste here: 

### Paste the markdown table here

| unique_id | age  | bp   | bgr   | bu   | sc   | sod   | pot   | hemo  | pcv  | pc      | pcc        | ba         | htn | dm | cad | appet | pe  | ane | Target  |
|-----------|------|------|-------|------|------|-------|-------|-------|------|---------|------------|------------|-----|----|-----|-------|-----|-----|---------|
| 203694    | 38   | 80   | 99    | 19   | 0.5  | 147   | 3.5   | 13.6  | 44   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 938027    | 43   | 60   | 108   | 25   | 1.0  | 144   | 5.0   | 17.8  | 43   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 938027    | 43   | 60   | 108   | 25   | 1.0  | 144   | 5.0   | 17.8  | 43   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 421471    | 37   | 60   | 111   | 35   | 0.8  | 135   | 4.1   | 16.2  | 50   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 109053    | 64   | 60   | 106   | 27   | 0.7  | 150   | 3.3   | 14.4  | 42   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 349892    | 59   | 70   | 424   | 55   | 1.7  | 138   | 4.5   | 12.6  | 37   | abnormal| notpresent | notpresent | yes | yes | yes | good  | no  | no  | ckd     |
| 630177    | 58   | 80   | 131   | 18   | 1.1  | 141   | 3.5   | 15.8  | 53   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 741276    | 23   | 60   | 95    | 24   | 0.8  | 145   | 5.0   | 15.0  | 52   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 546010    | 49   | 80   | 122   | 32   | 1.2  | 139   | 3.9   | 17.0  | 41   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 460465    | 47   | 60   | 109   | 25   | 1.1  | 141   | 4.7   | 15.8  | 41   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 965166    | 48   | 80   | 75    | 22   | 0.8  | 137   | 5.0   | 16.8  | 51   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 201677    | 33   | 80   | 89    | 19   | 1.1  | 144   | 5.0   | 15.0  | 40   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 976875    | 33   | 60   | 80    | 25   | 0.9  | 146   | 3.5   | 14.1  | 48   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 431871    | 57   | 60   | 132   | 18   | 1.1  | 150   | 4.7   | 15.4  | 42   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |
| 431871    | 57   | 60   | 132   | 18   | 1.1  | 150   | 4.7   | 15.4  | 42   | normal  | notpresent | notpresent | no  | no  | no  | good  | no  | no  | notckd  |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

You can mitigate hallucinations by breaking up data sets in smaller sets. You can also ensure that the data you are seeing matches the data you inputed for this instance I can check to make sure the LLM entered my data into the table correctly because I outputeted the code directly above it. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [168]:
# Code to rename all the columns in the dataset

abbreviation_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wc': 'White Blood Cell Count',
    'rc': 'Red Blood Cell Count',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'class': 'Class'
}

# Rename the columns using the mapping
df = combineddata_nomissing
#All the other comments are 383GPT comments, this is a <redacted> comment: I am using the version of the data from part 3.3 to show the full scope of the name changing code. 
df.rename(columns=abbreviation_mapping, inplace=True)

# Display the modified DataFrame
print(df)

     unique_id   Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0       203694  38.0            80.0                  99.0        19.0   
1       938027  43.0            60.0                 108.0        25.0   
2       938027  43.0            60.0                 108.0        25.0   
3       421471  37.0            60.0                 111.0        35.0   
9       109053  64.0            60.0                 106.0        27.0   
..         ...   ...             ...                   ...         ...   
467     486397  20.0            70.0                 123.0        44.0   
469     462892  36.0            80.0                  85.0        16.0   
478     475200  64.0           100.0                 163.0        54.0   
479     168242  48.0            60.0                 112.0        44.0   
480     474407  21.0            90.0                 107.0        40.0   

     Serum Creatinine  Sodium  Potassium  Hemoglobin  Packed Cell Volume  ...  \
0                 0.5   147.0 

C:\Users\sagea\AppData\Local\Temp\ipykernel_27284\3833719368.py:34: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.rename(columns=abbreviation_mapping, inplace=True)


#### MY CODE IS RUNNING FINE... THIS ERROR POPPED UP AFTER I COMPLETED THE ASSIGNMENT. 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [169]:
# Converted SQL to Pandas code
### Literally copy and pasted the code form 383GPT, I have no idea what any of this means or what the new data frame is meant to represent. 
import pandas as pd

# Assuming df is your DataFrame with data
df = pd.DataFrame({
    # Example data
    'Target': ['notckd', 'ckd', 'notckd', 'notckd', 'ckd'],  # Example target classes
    # Add other columns as necessary
})

# Group by 'Target' and count occurrences
result = df.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)

   Target  count
0     ckd      2
1  notckd      3


*Describe what the above code does here* 

It seems like what this code does is counts the amount of times an instance occurs in the data frame, and outputs that number in a count column. 